contributor-docs/adrs/adr-001-custom-elements.md

```diff
@@ -0,0 +1,115 @@
+
+### Extensibility
+Building behaviors in React Hooks gives us the ability to provide things like state and state change hooks to the consumer of the component. This allows the user to build on additional behaviors to the component based on the state or other variables provided to the component consumer. Doing the same with custom elements would require listening to events on the document[^3] and reacting to them. This is certainly do-able, but goes against some of the foundational principles of React (reacting to changes in the DOM vs changes in React state).
+
+### Organizational Overhead
+- GitHub’s custom elements are all managed in different repos which introduces more maintenance overhead.
+- Reacting to changes will take a bit more time, as we’ll need to wait for changes to be made to the custom elements until we can update Primer React Components (versus having behavior Hooks in PRC that we can update at our leisure).
+- Engineers who want to contribute to Primer React Components to build new components and behaviors would need to be familiar with both custom elements and React, two very different paradigms, and context switch between the two.
+
+### Other
+- The custom element & web component API progress slower than React due to changes needing to go through the whatwq standards process.
+
+### Risks of not switching to custom elements for behaviors
+- We spend extra time building behaviors in React that have already been built in our [custom elements library](https://github.github.io/web-systems-documentation/#custom-elements).
+  - There are currently 19 behaviors/components listed on the custom elements documentation site. Several of these we have already implemented in React in either PRC, Doctocat, or other React applications at GitHub which can be upstreamed (details-dialog, details-menu, clipboard-copy, text-expander, autocomplete, task-list via drag and drop hooks, tab-container, text-expander).
+- We decide not to invest further in React at GitHub and have wasted time we could have spent building more custom elements.
+  - This seems unlikely as there seems to be clear consensus that we will continue to build more and more highly interactive products.
+- The React library is abandoned and becomes obsolete
+  - This is a risk with any technology that we may use, seems highly unlikely in the near term, and is also a risk for custom elements.
```

Comment:
This isn't an existential risk to Custom Elements. Browsers can't deprecate API; the APIs will remain indefinitely.

Reply:
Yes, but it's possible that further work on them could be dropped

Reply:
From @keithamus in https://github.com/primer/components/pull/981#discussion_r564714720:
> …Browsers can't deprecate API…

Could you please clarify this, @keithamus—are you talking about [Custom Elements APIs](https://developer.mozilla.org/en-US/docs/Web/Web_Components/Using_custom_elements) specifically?

I’m not sure this holds as a general principle: For example, AppCache’s [MDN page](https://developer.mozilla.org/en-US/docs/Web/HTML/Using_the_application_cache) begins with the following warnings:

> **Do not use this _application cache_ feature!** It is [in the process of being removed from the Web platform](https://html.spec.whatwg.org/multipage/browsers.html#offline).
> 
> - From Firefox 84 it has been removed ([bug 1619673](https://bugzilla.mozilla.org/show_bug.cgi?id=1619673)). It is also planned for removal in Chomium 90, and is deprecated in Safari.

> 👎 **Deprecated**
> This feature is no longer recommended. Though some browsers might still support it, it may have already been removed from the relevant web standards, may be in the process of being dropped, or may only be kept for compatibility purposes. Avoid using it, and update existing code if possible; see the [compatibility table](https://developer.mozilla.org/en-US/docs/Web/HTML/Using_the_application_cache#Browser_compatibility) at the bottom of this page to guide your decision. Be aware that this feature may cease to work at any time.

Reply:
`AppCache` is an extremely rare case where browsers were able to deprecate an API. However it serves a good example of just how carefully APIs get deprecated, if ever. AppCache was a good candidate for deprecation as it was overly complicated, had serious architectural concerns, hampered development of new specs, not quite fit for purpose, and only 1.5% of the web used it. Most use cases used the declarative HTML attribute which meant disabling the API would have a minimal effect. Even so the API was originally marked for deprecation back in 2015, and it took several years for browsers to begin marking it as deprecated, with Chrome removing it from non-HTTPS sites in 2018, and shipping an origin trial removing it from HTTPS sites in 2020. As of writing Chrome still hasn’t deprecated AppCache - it’s due for removal in Chrome 90. In addition to this the Chrome team has provided a drop in library for those who are unable to remove the feature from their sites. So AppCache, a feature used by a small % of the web, has taken 6 years to get close to being deprecated, with browsers offering support for migrating throughout. 

Another example of a deprecated api would be Custom Elements v0 of which GitHub was a customer. This had a similar story: small amounts of usage, preventing new developments; with deprecation announced in 2014 via many deprecation notices, as well as support libraries offered. A paved path was provided to upgrade to the v1 spec and it took around 6 years to finally remove v0, with Chrome 80s release in 2020. GitHub migrated our existing components from v0 to v1, and Google engineers communicated with us to support the transition (see https://github.com/github/time-elements/pull/67 for an example). 

The difference between browsers deprecating APIs and, say, react deprecating APIs, is that the browsers have to maintain compatibility with websites. That's why we still have a `<font>` tag. React can release a major version and we either have to migrate or stick with an older version which may no longe receive support. The web can deprecate APIs but it’s going to be a multi year process and ultimately they’re beholden to web authors will to change.